[SPARK-14070] [SQL] Use ORC data source for SQL queries on ORC tables

tejasapatil opened the pull request:
## What changes were proposed in this pull request?

This patch enables use of OrcRelation for SQL queries which read data from Hive tables. Changes in this patch:
- Added a new rule `OrcConversions` which would alter the plan to use `OrcRelation`. In this diff, the conversion is done only for reads.
- Added a new config `spark.sql.hive.convertMetastoreOrc` to control the conversion

BEFORE 

```
scala>  hqlContext.sql("SELECT * FROM orc_table").explain(true)
== Parsed Logical Plan ==
'Project [unresolvedalias(*, None)]
+- 'UnresolvedRelation `orc_table`, None

== Analyzed Logical Plan ==
key: string, value: string
Project [key#171,value#172]
+- MetastoreRelation default, orc_table, None

== Optimized Logical Plan ==
MetastoreRelation default, orc_table, None

== Physical Plan ==
HiveTableScan [key#171,value#172], MetastoreRelation default, orc_table, None
```

AFTER

```
scala> hqlContext.sql("SELECT * FROM orc_table").explain(true)
== Parsed Logical Plan ==
'Project [unresolvedalias(*, None)]
+- 'UnresolvedRelation `orc_table`, None

== Analyzed Logical Plan ==
key: string, value: string
Project [key#76,value#77]
+- SubqueryAlias orc_table
   +- Relation[key#76,value#77] ORC part: struct<>, data: struct<key:string,value:string>

== Optimized Logical Plan ==
Relation[key#76,value#77] ORC part: struct<>, data: struct<key:string,value:string>

== Physical Plan ==
WholeStageCodegen
:  +- Scan ORC part: struct<>, data: struct<key:string,value:string>[key#76,value#77] InputPaths: file:/user/hive/warehouse/orc_table
```
## How was this patch tested?
- Added a new unit test. Ran existing unit tests
- Ran with production like data
## Performance gains

Ran on a production table in Facebook (note that the data was in DWRF file format which is similar to ORC)

Best case : when there was no matching rows for the predicate in the query (everything is filtered out)

```
                      CPU time          Wall time     Total wall time across all tasks
================================================================
Without the change   541_515 sec    25.0 mins    165.8 hours
With change              407 sec       1.5 mins     15 mins
```

Average case: A subset of rows in the data match the query predicate

```
                        CPU time        Wall time     Total wall time across all tasks
================================================================
Without the change   624_630 sec     31.0 mins    199.0 h
With change           14_769 sec      5.3 mins      7.7 h
```

tejasapatil committed:
Use ORC data source for SQL queries on ORC tables

AmplabJenkins commented:
Can one of the admins verify this patch?

marmbrus commented:
ok to test

SparkQA commented:
**[Test build #53806 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53806/consoleFull)** for PR 11891 at commit [`0938c08`](https://github.com/apache/spark/commit/0938c082cc1cb0b129f41340f1e75767aedbc3e1).

SparkQA commented:
**[Test build #53806 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53806/consoleFull)** for PR 11891 at commit [`0938c08`](https://github.com/apache/spark/commit/0938c082cc1cb0b129f41340f1e75767aedbc3e1).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53806/
Test FAILed.

rxin commented:
This is pretty cool -- the hive path is ridiculously slow. BTW I tried comparing Parquet vs ORC based on Spark master branch right now. 

I generated 100 million rows with one double column and one string column:

```
sqlContext.range(100 * 1000 * 1000)
  .select(rand().as("numeric"), rand().cast("string").as("string"))
  .write.parquet("testdata/random.parquet")
```

then read it back just the string column:

```
def measureParquet() {
  val start = System.nanoTime
  sqlContext.read.parquet("testdata/random.parquet").selectExpr("count(string)").show()
  val end = System.nanoTime
  print((end - start) / 1000 / 1000)
}
measureParquet()
```

Parquet with gzip compression takes ~12 secs. Parquet with snappy compression takes ~7 secs. ORC takes ~24 secs. We can definitely optimize the current ORC implementation more too.

tejasapatil commented:
@marmbrus : I looked at the build failures trying to figure out the cause : https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53806/ 

I need help in trying to just run the one of the tests that failed (preferably in Intellij). Do you know how to do that ?

tejasapatil commented:
@rxin : Can you point me to specific features / changes in Parquet which are not in ORC ? I am happy to work on adding that to ORC.

marmbrus commented:
I've never gotten testing running to work in intelij unfortunately.  You can use SBT as follows:

```
spark/$ sbt/sbt -Phive
> hive/test-only *HiveCompatibilitySuite -- -z date_serde
....
```

For the first failure, this looks like the cause:
```
[info]   Caused by: java.lang.UnsupportedOperationException: unsupported plan Relation[c1#282,c2#283] HadoopFiles
[info]   
[info]          at org.apache.spark.sql.hive.SQLBuilder.org$apache$spark$sql$hive$SQLBuilder$$toSQL(SQLBuilder.scala:191)
[info]          at org.apache.spark.sql.hive.SQLBuilder.org$apache$spark$sql$hive$SQLBuilder$$toSQL(SQLBuilder.scala:149)
[info]          at org.apache.spark.sql.hive.SQLBuilder.projectToSQL(SQLBuilder.scala:208)
[info]          at org.apache.spark.sql.hive.SQLBuilder.org$apache$spark$sql$hive$SQLBuilder$$toSQL(SQLBuilder.scala:111)
[info]          at org.apache.spark.sql.hive.SQLBuilder.org$apache$spark$sql$hive$SQLBuilder$$toSQL(SQLBuilder.scala:149)
[info]          at org.apache.spark.sql.hive.SQLBuilder.projectToSQL(SQLBuilder.scala:208)
[info]          at org.apache.spark.sql.hive.SQLBuilder.org$apache$spark$sql$hive$SQLBuilder$$toSQL(SQLBuilder.scala:111)
[info]          at org.apache.spark.sql.hive.SQLBuilder.toSQL(SQLBuilder.scala:81)
[info]          at org.apache.spark.sql.hive.execution.HiveComparisonTest$$anonfun$createQueryTest$1$$anonfun$36.liftedTree1$1(HiveComparisonTest.scala:398)
[info]          ... 57 more
[info]   
```

Seems we can't build SQL for ORC queries (used in view canonicalization) anymore.  @liancheng might have some suggestions.

marmbrus commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala:
```diff
@@ -597,6 +619,107 @@ private[hive] class HiveMetastoreCatalog(val client: HiveClient, hive: HiveConte
     }
   }
 
+  private def convertToOrcRelation(metastoreRelation: MetastoreRelation): LogicalRelation = {
```
How much of this is duplicated with `convertToParquetRelation`?  It would be awesome if we could just make this `ConvertToHadoopFSRelation` and handle all the cases (or just ORC/Parquet through a common path).  I'd have to look closer, but I think its likely that the parquet specific things (like parquetOption with the merged schema) are actually not required anymore.

The fact that this code path is parquet specific predates a common interface for reading files in Spark SQL.

liancheng replied:
Agree. It's probably OK to tolerant this duplication for this PR. I'd like to revisit this code path after finishing migrating ORC read path.

tejasapatil replied:
I have done a refac to have a single method named `convertToLogicalRelation` instead of separate ones for parquet and orc

liancheng commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/OrcQuerySuite.scala:
```diff
@@ -400,4 +400,33 @@ class OrcQuerySuite extends QueryTest with BeforeAndAfterAll with OrcTest {
+        }
+      }
+    }
+  }
```
Seems that this test case doesn't check whether the ORC table is actually read by our data source code path?

tejasapatil replied:
@liancheng : You are right. I had verified the plan generated by looking at it but was not able to capture the plan as a string and compare against expected one. On other hand, I felt that it will be a bad idea given that other changes to the system might have the expected plan changing which will make the test case cause of annoyance for dev. 

Any advice ?

liancheng replied:
Actually you can check the query plan. For example, we can check whether there is a `LocalRelation`:

```scala
val analyzedPlan = sqlContext
  .sql("SELECT * FROM dummy_orc WHERE key=0")
  .queryExecution
  .analyzedPlan

assert(analyzedPlan.collectFirst { case _: LocalRelation => () }.nonEmpty)
```

tejasapatil replied:
@liancheng : This is better than what I had before. Did the change.

liancheng commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/OrcQuerySuite.scala:
```diff
@@ -400,4 +400,33 @@ class OrcQuerySuite extends QueryTest with BeforeAndAfterAll with OrcTest {
+            sqlContext.sql(
+              s"""INSERT INTO TABLE dummy_orc
+                  |SELECT key, value FROM single
+               """.stripMargin).collect()
```
`INSERT` is a command, which is eagerly executed, so you don't need to `collect()` it.

tejasapatil replied:
removed `collect()`

liancheng commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala:
```diff
@@ -439,6 +440,56 @@ private[hive] class HiveMetastoreCatalog(val client: HiveClient, hive: HiveConte
+        expectedFileFormat match {
+          case `cachedRelationFileFormatClass` =>
+            // If we have the same paths, same schema, and same partition spec,
+            // we will use the cached relation.
```
And bucketing spec?

tejasapatil replied:
Done

tejasapatil replied:
done

liancheng commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala:
```diff
@@ -597,6 +619,107 @@ private[hive] class HiveMetastoreCatalog(val client: HiveClient, hive: HiveConte
+  /**
+    * When scanning Metastore ORC tables, convert them to ORC data source relations
+    * for better performance.
+    */
```
Nit: Please align beneath the first `*`.

tejasapatil replied:
done

liancheng commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala:
```diff
@@ -597,6 +619,107 @@ private[hive] class HiveMetastoreCatalog(val client: HiveClient, hive: HiveConte
+    * for better performance.
+    */
+  object OrcConversions extends Rule[LogicalPlan] {
+    private def isConvertMetastoreOrc(relation: MetastoreRelation): Boolean = {
```
`shouldConvertMetastoreOrc`?

tejasapatil replied:
done

tejasapatil committed:
Changes as per review comments (round #1)

- Merged the conversion methods for parquet and orc in one single method `convertToLogicalRelation` as suggested by @@marmbrus
- @liancheng's comment about checking for bucket spec
- 4 tests were failing because of the change. My change alters the plan for the queries.

eg. Query from `date_serde.q` :

```
select * from date_serde_orc
```

Plan (BEFORE)
```
Project [c1#282,c2#283]
+- MetastoreRelation default, date_serde_orc, None
```

Plan (AFTER)
```
Project [c1#287,c2#288]
+- SubqueryAlias date_serde_orc
   +- Relation[c1#287,c2#288] HadoopFiles
```

Setting `CONVERT_METASTORE_ORC` to `false` by default to mitigate test failures. Other option was to make `SQLBuilder` work with `Relation` but that is out of scope of the current PR. In my opinion, it would be better to have this config turned on by default so that anyone trying out Spark out of the box gets better perf. w/o needing to tweak such configs.

Open items:
- @liancheng's review comment : Test case added does not verify if the new codepath is hit

SparkQA commented:
**[Test build #53968 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53968/consoleFull)** for PR 11891 at commit [`c71bb6c`](https://github.com/apache/spark/commit/c71bb6c1a06de3905a0b1103260073703d56a968).

SparkQA commented:
**[Test build #53968 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53968/consoleFull)** for PR 11891 at commit [`c71bb6c`](https://github.com/apache/spark/commit/c71bb6c1a06de3905a0b1103260073703d56a968).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53968/
Test FAILed.

tejasapatil committed:
Fixed checkstyle

SparkQA commented:
**[Test build #53969 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53969/consoleFull)** for PR 11891 at commit [`c15a39f`](https://github.com/apache/spark/commit/c15a39f4963a7c6481c667d059579e628763a867).

SparkQA commented:
**[Test build #53969 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53969/consoleFull)** for PR 11891 at commit [`c15a39f`](https://github.com/apache/spark/commit/c15a39f4963a7c6481c667d059579e628763a867).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53969/
Test FAILed.

tejasapatil commented:
@liancheng + @marmbrus : Thanks for your comments. I have made the suggested changes except the one related to test case which I am not sure how to do.

Also, the PR failed to build on Jenkins but it builds fine on my end when I do:

  build/sbt scalastyle catalyst/compile sql/compile hive/compile
  build/sbt -Pyarn -Phadoop-2.3 -Phive package assembly/assembly

I looked at the jenkins log and I *guess* that incremental compilation makes it not see the new class `OrcConversions` added in `HiveSessionCatalog`. Is there a way to avoid that ?

tejasapatil committed:
Setting `CONVERT_METASTORE_ORC` to `true` by default. This means that anyone downloading Spark and trying it out will get the best performance w/o needing to know about this feature and tweak this config.

For the tests which had failed, made `HiveCompatibilitySuite` to set that config to `false` so that they still produce the plan as per the old way.

Ran `build/sbt scalastyle catalyst/compile sql/compile hive/compile`
Ran `build/sbt -Phive hive/test-only *HiveCompatibilitySuite`

SparkQA commented:
**[Test build #53983 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53983/consoleFull)** for PR 11891 at commit [`c4111fa`](https://github.com/apache/spark/commit/c4111fa9ea5b18bc2efa02e6f3da3739e6a5f669).

SparkQA commented:
**[Test build #53983 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/53983/consoleFull)** for PR 11891 at commit [`c4111fa`](https://github.com/apache/spark/commit/c4111fa9ea5b18bc2efa02e6f3da3739e6a5f669).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/53983/
Test FAILed.

liancheng commented:
@tejasapatil When testing a PR, Jenkins always tries to merge it with the most recent master (or any other branch against which the PR is opened) first. I think the reason of the build failure is that we just merged PR #11836, which probably doesn't conflict with yours from Git's point of view but actually causes compilation failure. Would you please try to rebase to the most recent master and retry?

liancheng commented:
@tejasapatil Just saw your previous comment about debugging in IntelliJ. When I have to resort to the debugger, I usually run tests using SBT and debug with IntelliJ remote debugging feature. [This][1] is probably useful for you. Under SBT, you may do this:

```
> project hive // or switch to any other project the target test case belongs to
> set javaOptions in Test += <remote debugging options>
> test-only ...
...
> session clear // this clears the previously added options and disable remote debugging
```

[1]: https://mail-archives.apache.org/mod_mbox/spark-dev/201504.mbox/<551FDD23.3010407@gmail.com>

tejasapatil commented:
@liancheng : I rebased and fixed the problem. Also, thanks for the pointer for debugging the tests.

tejasapatil commented:
ok to test

liancheng commented:
retest this please

liancheng commented:
add to whitelist

SparkQA commented:
**[Test build #54010 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54010/consoleFull)** for PR 11891 at commit [`c4111fa`](https://github.com/apache/spark/commit/c4111fa9ea5b18bc2efa02e6f3da3739e6a5f669).

SparkQA commented:
**[Test build #54010 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54010/consoleFull)** for PR 11891 at commit [`c4111fa`](https://github.com/apache/spark/commit/c4111fa9ea5b18bc2efa02e6f3da3739e6a5f669).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54010/
Test FAILed.

liancheng commented:
Ah, unfortunately we just reverted #11836 because it caused some regressions. This caused another compilation error for your PR... Sorry for the trouble...

SparkQA commented:
**[Test build #54071 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54071/consoleFull)** for PR 11891 at commit [`0a03a78`](https://github.com/apache/spark/commit/0a03a7882b2d5edf972b1976b4fc223052909a7c).

SparkQA commented:
**[Test build #54071 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54071/consoleFull)** for PR 11891 at commit [`0a03a78`](https://github.com/apache/spark/commit/0a03a7882b2d5edf972b1976b4fc223052909a7c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54071/
Test PASSed.

tejasapatil committed:
dummy commit

tejasapatil committed:
revert dummy commit

SparkQA commented:
**[Test build #54092 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54092/consoleFull)** for PR 11891 at commit [`4b459a0`](https://github.com/apache/spark/commit/4b459a076ddd9ac95a66a695dc313a13253b1fd4).

tejasapatil committed:
Modified the test case as per @liancheng's suggestion

SparkQA commented:
**[Test build #54097 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54097/consoleFull)** for PR 11891 at commit [`4f0fe06`](https://github.com/apache/spark/commit/4f0fe065279edbffeba70609fde837e7dfa17e58).

SparkQA commented:
**[Test build #54092 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54092/consoleFull)** for PR 11891 at commit [`4b459a0`](https://github.com/apache/spark/commit/4b459a076ddd9ac95a66a695dc313a13253b1fd4).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54092/
Test PASSed.

SparkQA commented:
**[Test build #54097 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54097/consoleFull)** for PR 11891 at commit [`4f0fe06`](https://github.com/apache/spark/commit/4f0fe065279edbffeba70609fde837e7dfa17e58).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54097/
Test PASSed.

tejasapatil committed:
rebased

tejasapatil commented:
ok to test

SparkQA commented:
**[Test build #54267 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54267/consoleFull)** for PR 11891 at commit [`3c25e7e`](https://github.com/apache/spark/commit/3c25e7ec503f3bc705d2635a13d235bf6e0ec5a1).

SparkQA commented:
**[Test build #54267 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/54267/consoleFull)** for PR 11891 at commit [`3c25e7e`](https://github.com/apache/spark/commit/3c25e7ec503f3bc705d2635a13d235bf6e0ec5a1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/54267/
Test PASSed.

tejasapatil commented:
@liancheng : I have made all requested changes as per review and also rebased. Can you please take a look ?

tejasapatil commented:
ping @liancheng

marmbrus commented:
Thanks, merging to master.
